### PR TITLE
Retry SNS event if fails

### PR DESCRIPTION
**Pull request by garethhallberg**
Retry using spring retry. Async done with Coroutines. Feature is off for now. Can be turned on by changing values in application=local.yml

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -308,13 +309,14 @@ tasks.get("openApiGenerate").doLast {
...
-      val replacedFileContents = FileUtils.readFileToString(it, "UTF-8")
-        .replace("include = JsonTypeInfo.As.PROPERTY", "include = JsonTypeInfo.As.EXISTING_PROPERTY")
-      FileUtils.writeStringToFile(it, replacedFileContents, "UTF-8")
+  File("$rootDir/build/generated/src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/api/model").walk()
+    .forEach {
+      if (it.isFile && it.extension == "kt") {
+        val replacedFileContents = FileUtils.readFileToString(it, "UTF-8")
+          .replace("include = JsonTypeInfo.As.PROPERTY", "include = JsonTypeInfo.As.EXISTING_PROPERTY")
+        FileUtils.writeStringToFile(it, replacedFileContents, "UTF-8")
+      }
```
Comment (pezholio) [inline build.gradle.kts:318]: Are these changes autolinting things?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt [hunk 2]
```diff
@@ -207,35 +213,130 @@ class DomainEventService(
...
+        typeName,
+        typeDescription,
+        detailUrl,
+        crn,
+        nomsNumber,
+      )
+    }
+  }
+
+  final fun retryTemplate(): RetryTemplate {
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:242]: I think it'd be better to have this at the bottom of the class to save misdirection

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt [hunk 3]
```diff
@@ -207,35 +213,130 @@ class DomainEventService(
...
+
+  fun doAsyncSave(
+    domainEvent: DomainEvent<*>,
+    typeName: String,
+    typeDescription: String,
+    detailUrl: String,
+    crn: String,
+    nomsNumber: String,
+  ) = runBlocking {
+    val snsEvent = SnsEvent(
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:261]: Can we add shared functions to generate the `snsEvent` and  `publishRequest`, as they're essentially the same in `doAsyncSave` and `doSyncSave`?

**Review by pezholio — COMMENTED** [review]
Thanks for persevering with this! Just a couple of comments to tighten things up

File: src/main/resources/application-local.yml [hunk 4]
```diff
@@ -22,10 +22,13 @@ hmpps.sqs:
 domain-events:
   cas1:
     emit-enabled: true
+    async-save-enabled: false
   cas2:
     emit-enabled: true
+    async-save-enabled: false
   cas3:
-    emit-enabled: bookingCancelled,bookingConfirmed,bookingProvisionallyMade,personArrived,personDeparted,referralSubmitted,personDepartureUpdated,bookingCancelledUpdated,personArrivedUpdated
+    emit-enabled: bookingCancelled,bookingConfirmed,bookingProvisionallyMade,personArrived,personDeparted,referralSubmitted,personDepartureUpdated,bookingCancelledUpdated
```
[the remark this answers is not in the document]
Comment (garethhallberg) [reply src/main/resources/application-local.yml:30]: yes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt [hunk 5]
```diff
@@ -207,35 +213,125 @@ class DomainEventService(
...
+    val snsEvent = createSnsEvent(
+      domainEvent,
+      typeName,
+      typeDescription,
+      detailUrl,
+      crn,
+      nomsNumber,
+    )
+
+    val publishRequest = createPublishRequest(snsEvent)
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:291]: Could probs improve this further by creating the `snsEvent` and the `publishRequest` in `saveAndEmit`, which would make both methods a bit simpler.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt [hunk 6]
```diff
@@ -207,35 +213,125 @@ class DomainEventService(
...
+
+    retry(publishRequest, domainEvent, snsEvent)
+  }
+
+  suspend fun retry(publishRequest: PublishRequest, domainEvent: DomainEvent<*>, snsEvent: SnsEvent) {
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:296]: Might be better to give this a more descriptive name, something like `publishSnsEventWithRetry`

**Review by pezholio — COMMENTED** [review]
Just a coupla further comments to tighten this up even more 😄

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt [hunk 7]
```diff
@@ -249,4 +345,20 @@ class DomainEventService(
...
+    backOffPolicy.backOffPeriod = BACK_OFF_PERIOD
+    retryTemplate.setBackOffPolicy(backOffPolicy)
+    val retryPolicy = SimpleRetryPolicy()
+    retryPolicy.maxAttempts = MAX_ATTEMPTS_RETRY
+    retryTemplate.setRetryPolicy(retryPolicy)
+    return retryTemplate
+  }
+
+  companion object {
+    const val MAX_ATTEMPTS_RETRY = 4
```
Comment (tahb) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:361]: What do we think about 4 here? I think if these requests ever fail something serious has gone wrong that may persist for 12h or 24h. I worry about us releasing a change that causes failures and by the time we notice and fix it the queue has given up. How about the equivalent of 30 days?

Comment (garethhallberg) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:361]: Keep trying every 1000 milliseconds for 30 days?

Comment (tahb) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:361]: I'd assume an increasing backoff though judging our caching I presume we'd have to figure that bit out?

Every second is too often 😅 

Perhaps 4 at 1 minute intervals, 4 at 10 minutes intervals and then however many an hour to give us a big window for it to automatically recover.

Comment (garethhallberg) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:361]: OK I'll look into how we do this with retry template or if it's even possible.

Comment (garethhallberg) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt:372]: @tahb I've used the ExponentialBackOffPolicy with it's default multiplier of 2. So we start off with 1 second then 2 second then 4 second etc until we get to ten minutes. We stick with ten minutes for the next seven days. Discussed with Stu that seven days should be adequate.

**Comment by pezholio** [comment]
Sorry for being a nitpicker, but I've been thinking about this and the service is beginning to look a bit crowded now. I'm wondering if we can borrow from @danlivings-dxw's recent approach with the data sources and add a new configurable worker class - I've sketched this out and have no idea if it works, but it seems to compile!

```kotlin
package uk.gov.justice.digital.hmpps.approvedpremisesapi.workers

import com.amazonaws.services.sns.model.MessageAttributeValue
import com.amazonaws.services.sns.model.PublishRequest
import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.coroutines.runBlocking
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Primary
import org.springframework.beans.factory.annotation.Value
import org.springframework.retry.support.RetryTemplate
import org.springframework.stereotype.Component
import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.DomainEvent
import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEvent
import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventAdditionalInformation
import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventPersonReference
import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventPersonReferenceCollection
import uk.gov.justice.hmpps.sqs.HmppsQueueService
import uk.gov.justice.hmpps.sqs.MissingTopicException
import java.time.ZoneOffset

interface DomainEventWorkerInterface {
  fun emitEvent(options: DomainEventWorkerOptions): Unit
}

@Component
@Primary
class ConfiguredDomainEventWorker(
  private val objectMapper: ObjectMapper,
  private val hmppsQueueService: HmppsQueueService,
  @Value("\${domain-events.cas1.async-save-enabled}") private val asyncSaveEnabled: Boolean,
) : DomainEventWorkerInterface {

  private var worker: DomainEventWorkerInterface = if (asyncSaveEnabled) {
    AsyncDomainEventWorker(objectMapper, hmppsQueueService)
  } else {
    SyncDomainEventWorker(objectMapper, hmppsQueueService)
  }

  override fun emitEvent(options: DomainEventWorkerOptions) = this.worker.emitEvent(options)
}

@Component
class SyncDomainEventWorker(
  private val objectMapper: ObjectMapper,
  private val hmppsQueueService: HmppsQueueService,
) : DomainEventWorker(objectMapper, hmppsQueueService) {
  override fun emitEvent(options: DomainEventWorkerOptions) {
    val snsEvent = createSnsEvent(options)
    val publishRequest = createPublishRequest(snsEvent)
    val publishResult = domainTopic.snsClient.publish(publishRequest)

    log.info(
      "Emitted SNS event (Message Id: ${publishResult.messageId}, " +
        "Sequence Id: ${publishResult.sequenceNumber}) for Domain Event:" +
        " ${options.domainEvent.id} of type: ${snsEvent.eventType}",
    )
  }
}

@Component
class AsyncDomainEventWorker(
  private val objectMapper: ObjectMapper,
  private val hmppsQueueService: HmppsQueueService,
  ) : DomainEventWorker(objectMapper, hmppsQueueService) {
    // This allows us to inject a new retryTemplate in the tests if we want
    var retryTemplate: RetryTemplate = RetryTemplate.builder()
      .exponentialBackoff(INITIAL_INTERVAL, 2.0, MAX_INTERVAL)
      .maxAttempts(MAX_ATTEMPTS_RETRY_EXPONENTIAL)
      .build()

  override fun emitEvent(options: DomainEventWorkerOptions) {
      val snsEvent = createSnsEvent(options)
      val publishRequest = createPublishRequest(snsEvent)
      runBlocking {
        publishSnsEventWithRetry(publishRequest, options.domainEvent, snsEvent, retryTemplate)
      }
    }

  suspend fun publishSnsEventWithRetry(
    publishRequest: PublishRequest,
    domainEvent: DomainEvent<*>,
    snsEvent: SnsEvent,
    retryTemplate: RetryTemplate,
  ) {
    retryTemplate.execute<Any?, RuntimeException> {
      val publishResult = domainTopic.snsClient.publish(publishRequest)
      log.info(
        "Emitted SNS event (Message Id: ${publishResult.messageId}, " +
          "Sequence Id: ${publishResult.sequenceNumber}) for Domain Event: " +
          "${domainEvent.id} of type: ${snsEvent.eventType}",
      )
    }
  }

  companion object {
    /*
    * The MAX_ATTEMPTS_RETRY_EXPONENTIAL of 166 is worked out like this:
    * the multiplier is the default of two so we have
    * 1 2 4 8 16 32 64 128 256 512 which gives us around 25 minutes. After that our max interval is ten minutes
    * so we have four of them to give us the first hour which is 14 attempts. We follow that with 23 * 6 for the rest
    * of the fist day which gives us 14 + 138 = 152. We want to retry for 7 days so we have another 6 * 6 * 24 = 864.
    * Total is 864 + 152 = 1016
    *
    * MAX_ATTEMPTS_RETRY is used in the tests with a FixedBackOffPolicy
    * */
    const val MAX_ATTEMPTS_RETRY_EXPONENTIAL = 1016
    const val MAX_ATTEMPTS_RETRY = 4
    const val BACK_OFF_PERIOD = 1000.toLong()
    const val INITIAL_INTERVAL = 1000.toLong()
    const val MAX_INTERVAL = (1000 * 60 * 10).toLong()
  }
}

open class DomainEventWorker(
  private val objectMapper: ObjectMapper,
  private val hmppsQueueService: HmppsQueueService,
) : DomainEventWorkerInterface {
  val log = LoggerFactory.getLogger(this::class.java)
  val domainTopic by lazy {
    hmppsQueueService.findByTopicId("domainevents")
      ?: throw MissingTopicException("domainevents not found")
  }

  fun createSnsEvent(options: DomainEventWorkerOptions): SnsEvent = SnsEvent(
    eventType = options.typeName,
    version = 1,
    description = options.typeDescription,
    detailUrl = options.detailUrl,
    occurredAt = options.domainEvent.occurredAt.atOffset(ZoneOffset.UTC),
    additionalInformation = SnsEventAdditionalInformation(
      applicationId = options.domainEvent.applicationId,
    ),
    personReference = SnsEventPersonReferenceCollection(
      identifiers = listOf(
        SnsEventPersonReference("CRN", options.crn),
        SnsEventPersonReference("NOMS", options.nomsNumber),
      ),
    ),
  )

  fun createPublishRequest(snsEvent: SnsEvent): PublishRequest =
    PublishRequest(domainTopic.arn, objectMapper.writeValueAsString(snsEvent))
      .withMessageAttributes(
        mapOf(
          "eventType" to MessageAttributeValue().withDataType("String").withStringValue(snsEvent.eventType),
        ),
      )

  override fun emitEvent(options: DomainEventWorkerOptions) {
    throw NotImplementedError("Cannot call emitEvent from the DomainEventWorker class")
  }
}

// Create a data class to make the arguments we pass around tider (and keep Detekt happy!)
data class DomainEventWorkerOptions(
  val domainEvent: DomainEvent<*>,
  val typeName: String,
  val typeDescription: String,
  val detailUrl: String,
  val crn: String,
  val nomsNumber: String,
)
```

Then in the service we can add the `configuredDomainEventWorker` to the service like so:

```kotlin
@Service
class DomainEventService(
  private val objectMapper: ObjectMapper,
  private val domainEventRepository: DomainEventRepository,
  private val hmppsQueueService: HmppsQueueService,
  private val domainEventWorker: ConfiguredDomainEventWorker,
  @Value("\${domain-events.cas1.emit-enabled}") private val emitDomainEventsEnabled: Boolean,
  @Value("\${domain-events.cas1.async-save-enabled}") private val asyncSaveEnabled: Boolean,
  @Value("\${url-templates.api.cas1.application-submitted-event-detail}") private val applicationSubmittedDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.application-assessed-event-detail}") private val applicationAssessedDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.booking-made-event-detail}") private val bookingMadeDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.person-arrived-event-detail}") private val personArrivedDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.person-not-arrived-event-detail}") private val personNotArrivedDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.person-departed-event-detail}") private val personDepartedDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.booking-not-made-event-detail}") private val bookingNotMadeDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.booking-cancelled-event-detail}") private val bookingCancelledDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.booking-changed-event-detail}") private val bookingChangedDetailUrlTemplate: String,
  @Value("\${url-templates.api.cas1.application-withdrawn-event-detail}") private val applicationWithdrawnDetailUrlTemplate: String,
) {
....
```

And call the worker like so

```kotlin
private fun saveAndEmit(
    domainEvent: DomainEvent<*>,
    typeName: String,
    typeDescription: String,
    detailUrl: String,
    crn: String,
    nomsNumber: String,
  ) {
    domainEventRepository.save(
      DomainEventEntity(
        id = domainEvent.id,
        applicationId = domainEvent.applicationId,
        bookingId = null,
        crn = domainEvent.crn,
        type = enumTypeFromDataType(domainEvent.data!!::class.java),
        occurredAt = domainEvent.occurredAt.atOffset(ZoneOffset.UTC),
        createdAt = OffsetDateTime.now(),
        data = objectMapper.writeValueAsString(domainEvent.data),
        service = "CAS1",
      ),
    )

    if (!emitDomainEventsEnabled) {
      log.info("Not emitting SNS event for domain event because domain-events.cas1.emit-enabled is not enabled")
      return
    }

    val options = DomainEventWorkerOptions(
      domainEvent,
      typeName,
      typeDescription,
      detailUrl,
      crn,
      nomsNumber,
    )

    domainEventWorker.emitEvent(options)
  }
```

That way we can keep the service tider, seperate the concerns a little better and we can unit test the worker stuff much more easily.

Happy to pair with you on this tomorrow if you fancy 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventWorkerTest.kt [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+  }
+  private val hmppsTopicMock = mockk<HmppsTopic>()
+
+  private val asyncDomainEventWorker = AsyncDomainEventWorker(
+    domainTopic = hmppsTopicMock,
+  )
+
+  private val syncDomainEventWorker = SyncDomainEventWorker(
+    domainTopic = hmppsTopicMock,
+  )
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventWorkerTest.kt:54]: I think it'd be worth initializing the `asyncDomainEventWorker` and `syncDomainEventWorker` using the `ConfiguredDomainEventWorker` interface, so we can test that works as expected:

```
private val asyncDomainEventWorker = ConfiguredDomainEventWorker(
  hmppsQueueService,
  false
)

private val syncDomainEventWorker = ConfiguredDomainEventWorker(
  hmppsQueueService,
  true
)
```

We could then mock the `hmppsQueueServiceMock` in a beforeEach like so:

```
@BeforeEach
fun setup() {
   every { hmppsQueueServieMock.findByTopicId("domainevents") } returns asyncDomainEventWorker.domainTopic
}
```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventWorkerTest.kt [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+import uk.gov.justice.hmpps.sqs.HmppsTopic
+import java.time.Instant
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import java.util.UUID
+
+@Suppress("SwallowedException")
+class DomainEventWorkerTest {
+  private val domainEventRespositoryMock = mockk<DomainEventRepository>()
+  private val hmppsQueueServieMock = mockk<HmppsQueueService>()
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventWorkerTest.kt:40]: Typo here 😃 

```suggestion
  private val hmppsQueueServiceMock = mockk<HmppsQueueService>()
```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventWorkerTest.kt [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+  @Test
+  fun `async emit retries if it fails`() {
+    val id = UUID.fromString("c3b98c67-065a-408d-abea-a252f1d70981")
+    val applicationId = UUID.fromString("a831ead2-31ae-4907-8e1c-cad74cb9667b")
+    val occurredAt = OffsetDateTime.parse("2023-02-01T14:03:00+00:00")
+    val crn = "CRN"
+
+    every { domainEventRespositoryMock.save(any()) } answers { it.invocation.args[0] as DomainEventEntity }
+    every { domainEventRespositoryMock.save(any()) } answers { it.invocation.args[0] as DomainEventEntity }
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventWorkerTest.kt:64]: We don't nee these mocks here

**Review by pezholio — COMMENTED** [review]
Looking good! Just a few relatively minor things, then I think we're good to go! 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventServiceTest.kt [hunk 11]
```diff
@@ -127,6 +131,7 @@ class DomainEventServiceTest {
     val mockHmppsTopic = mockk<HmppsTopic>()
 
     every { hmppsQueueServieMock.findByTopicId("domainevents") } returns mockHmppsTopic
+    every { domainEventService.domainEventWorker.emitEvent(any(), any(), any()) } returns Unit
```
[the remark this answers is not in the document]
Comment (pezholio) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventServiceTest.kt:134]: Sorry, should be:

```
every { domainEventWorkerMock.emitEvent(any(), any(), any()) } returns Unit
```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventServiceTest.kt [hunk 12]
```diff
@@ -36,18 +37,20 @@ import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.DomainEventEn
...
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.ConfiguredDomainEventWorker
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.DomainEventService
 import uk.gov.justice.hmpps.sqs.HmppsQueueService
 import uk.gov.justice.hmpps.sqs.HmppsTopic
 import java.time.Instant
 import java.time.OffsetDateTime
 import java.util.UUID
 
+// we are suppressing the exception as we test a RuntimeException gets thrown
+
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventServiceTest.kt:49]: I don't think we need this comment now

**Review by pezholio — APPROVED** [review]
LGTM - thanks for bearing with me! 😅 

I have just noticed one potential improvement, but I'll push it up once this is merged - it's not essential and I'm sure you want to work on something else now! 😆

**Pull request merged**